src/commcare_cloud/commands/terraform/aws.py

```diff
@@ -150,3 +159,116 @@ def run(self, args, unknown_args):
+
+    arguments = [
+        Argument('--duration-minutes', type=int, default=DEFAULT_SIGN_IN_DURATION_MINUTES, help="""
+            Stay signed in for this many minutes
+        """)
+    ]
+
+    def run(self, args, unknown_args):
+        environment = get_environment(args.env_name)
+        duration_minutes = args.duration_minutes
+        aws_profile = environment.terraform_config.aws_profile
+        aws_sign_in(aws_profile, duration_minutes, force_new=True)
+
+
+@memoized
+def aws_sign_in(aws_profile, duration_minutes=DEFAULT_SIGN_IN_DURATION_MINUTES,
+                force_new=False):
+    """
+    Create a temp session through MFA for a given aws profile
+
+    :param aws_profile: The name of an existing aws profile to create a temp session for
+    :param duration_minutes: How long to set the session expiration if a new one is created
+    :param force_new: If set to True, creates new credentials even if valid ones are found
+    :return: The name of temp session profile.
+             (Always the passed in profile followed by ':session')
+    """
+    aws_session_profile = '{}:session'.format(aws_profile)
+    if not force_new \
+            and _has_valid_session_credentials(aws_session_profile):
+        return aws_session_profile
+
+    default_username = getpass.getuser()
+    username = input("Enter username associated with credentials [{}]: ".format(
+        default_username)) or default_username
+    mfa_token = input("Enter your MFA token: ")
+    generate_session_profile(aws_profile, username, mfa_token, duration_minutes)
+
+    puts(colored.green(u"✓ Sign in accepted"))
+    puts(colored.cyan(
+        "You will be able to use AWS from the command line for the next {} minutes."
+        .format(duration_minutes)))
+    puts(colored.cyan(
+        "To use this session outside of commcare-cloud, "
+        "prefix your command with AWS_PROFILE={}:session".format(aws_profile)))
+    return aws_session_profile
+
+
+def generate_session_profile(aws_profile, username, mfa_token, duration_minutes):
+    # General idea from https://www.vividcortex.com/blog/setting-up-multi-factor-authentication-with-the-aws-cli
+    boto_session = boto3.session.Session(profile_name=aws_profile)
+    iam = boto_session.client('iam')
+    devices = iam.list_mfa_devices(UserName=username)['MFADevices']
+    device_arn = sorted(devices, key=lambda device: device['EnableDate'])[-1]['SerialNumber']
+    sts = boto_session.client('sts')
+    credentials = sts.get_session_token(
+        SerialNumber=device_arn,
+        TokenCode=mfa_token,
+        DurationSeconds=duration_minutes * 60,
+    )['Credentials']
+    _write_credentials_to_aws_credentials(
+        '{}:session'.format(aws_profile),
+        aws_access_key_id=credentials['AccessKeyId'],
+        aws_secret_access_key=credentials['SecretAccessKey'],
+        aws_session_token=credentials['SessionToken'],
+        expiration=credentials['Expiration'],
```

Comment:
Suggestion: pass in `credentials` instead of splitting out all these arguments, making `_write_credentials_to_aws_credentials` the only piece of code that knows the structure of this data and how it is persisted.

Reply:
I think like I like the way it is better myself, where I keep parsing contained to the place that made the API error, and then let `_write_credentials_to_aws_credentials` have an explicit interface; having a `_write_credentials_to_aws_credentials` take as an argument something that looks like the result of an AWS STS API call I don't like as much. (Imagine another place that already had these things calling it and having to call it like `_write_credentials_to_aws_credentials({'AccessKeyId': aws_access_key_id, ...})`; not that that's likely to happen.) Anyway, I think doing it the way you suggested would be fine too, but I think I'm going to keep it the way it is.